### PR TITLE
Max leverage checks

### PR DESCRIPTION
**Pull request by gmxer**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206849254647553
  - https://app.asana.com/0/0/1206579977668991
  - https://app.asana.com/0/0/1206215823771793
  - https://app.asana.com/0/0/1206939352715150

### CODE REVIEW

File: src/lib/useDebouncedInputValue.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  useEffect(() => {
+    if (timeoutRef.current) {
+      clearTimeout(timeoutRef.current);
+    }
+
+    timeoutRef.current = window.setTimeout(() => {
+      commit(value);
+    }, DEBOUNCE_MS);
+  });
```
Comment (onemicky) [inline src/lib/useDebouncedInputValue.ts:19]: Correct me if I'm wrong, but I think there may be a potential problem here due to the lack of deps array. In the event of permanent rerendering, we could get stuck in a situation where the commit never triggers.

File: src/components/Synthetics/OrderEditor/OrderEditor.tsx [hunk 2]
```diff
@@ -260,21 +266,40 @@ export function OrderEditor(p: Props) {
...
+      positionOrder?.isLong,
+      positionOrder?.marketAddress,
+      positionOrder?.targetCollateralToken.address,
+      triggerPrice,
+    ]
+  );
+  const nextPositionValuesForIncrease = useNextPositionValuesForIncrease(nextPositionValuesArg);
+  const nextPositionValuesWithoutPnlForIncrease = useNextPositionValuesForIncrease({
+    ...nextPositionValuesArg,
+    overrideIsPnlInLeverage: false,
```
[the remark this answers is not in the document]
Comment (gmxer) [reply src/components/Synthetics/OrderEditor/OrderEditor.tsx:302]: I destruct all the parameters in the `useNextPositionValuesForIncrease` so it wonn't change behavior

File: src/components/Synthetics/OrderEditor/OrderEditor.tsx [hunk 3]
```diff
@@ -457,8 +483,121 @@ export function OrderEditor(p: Props) {
...
+          collateralToken,
+          findSwapPath: swapRoute.findSwapPath,
+          indexToken: positionIndexToken,
+          indexTokenAmount,
+          initialCollateralAmount: positionOrder.initialCollateralDeltaAmount,
+          initialCollateralToken: fromToken,
+          isLong: positionOrder.isLong,
+          marketInfo: positionOrder.marketInfo,
+          position: existingPosition,
+          strategy: "leverageByCollateral",
```
[the remark this answers is not in the document]
Comment (gmxer) [reply src/components/Synthetics/OrderEditor/OrderEditor.tsx:528]: as I understand it doesn't exactly what it's supposed by it's name
as I understand
* `leverageByCollateral` - desired leverage and collateral amount are fixed, size is calculated based on lev and collateral.
* `leverageBySize` - desired leverage and position size are fixed, collateral is calculated, collateral is calculated based on lev and size.
* `independent` - leverage is caculated based on provided size and collateral

In all of this three calculations we only work with "desired" leverage (`estimatedLeverage`) inside `getIncreasePositionAmounts` fn. 

But the actual leverage is calculated after one of this strategy applied, and called `nextLeverage` in `getNextPositionValuesForIncreaseTrade`.

And in all of my cases I want to reestimate position size, keeping the leverage and collateral as user inputed.

Hope that makes sense, tricky place

File: src/lib/useDebouncedInputValue.ts (hunk 1 above)
Comment (gmxer) [reply src/lib/useDebouncedInputValue.ts:19]: wow, that's a catch

File: src/components/Synthetics/OrderEditor/OrderEditor.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (onemicky) [reply src/components/Synthetics/OrderEditor/OrderEditor.tsx:302]: Oh, that's right. Then useMemo for nextPositionValuesArg itself also can be omitted right?

File: src/components/Synthetics/OrderEditor/OrderEditor.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (onemicky) [reply src/components/Synthetics/OrderEditor/OrderEditor.tsx:528]: Got that! Thanks for the breakdown

**Review by onemicky — APPROVED** [review]

**Pull request merged**